Add support for Firefox and Edge

GeeTransit opened the pull request:
I tested Firefox and it works all fine. Not sure about Safari though. The `TyperacerBot` class is removed and its methods are now functions. There could be more prompts for configuration later on.

GeeTransit committed:
Add Firefox driver (geckodriver.exe)

GeeTransit committed:
Don't start python in new window

GeeTransit committed:
Unpack TyperacerBot class into functions

__init__ -> main
start -> driver.get(link)
find_text -> find_text(driver)
keypress_type -> keypress_type(keyboard, text, delay)
keysend_type -> keysend_type(driver, text, delay)

GeeTransit committed:
Fix error (Use keyboard, not driver)

GeeTransit committed:
Move link and delay into main's arguments

GeeTransit committed:
Add support for Firefox and Safari (not tested)

The `get_driver` function first checks the preferred driver (str), then checks all drivers in DRIVER_FUNCS. It raises LookupError on failure.

GeeTransit committed:
Rename key*_type -> *_text

GeeTransit committed:
Fix error (WebDriverException)

GeeTransit committed:
Remove try-except in send_text (let caller handle)

GeeTransit committed:
Add docstrings and general cleanup

GeeTransit committed:
Fix and simplify imports

Togohogo1 commented on run.bat:
```diff
@@ -7,4 +7,4 @@ echo installing requirements
 
 echo starting app
-start python typeracer_bot.py
+python typeracer_bot.py
```
kinda messy opening in the same window

Togohogo1 commented on typeracer_bot.py:
```diff
@@ -1,78 +1,81 @@
+        elem.send_keys(i)
+
+# dict[name: lambda -> WebDriver]
+DRIVER_FUNCS = dict(
```
lolz imma move this to the top

kwangzera commented on typeracer_bot.py:
```diff
@@ -1,78 +1,81 @@
+    safari=lambda: Safari(executable_path="driver/safaridriver"),
+)
+
+def get_driver(preferred=None, drivers=DRIVER_FUNCS):
```
Why not define `DRIVER_FUNCS` in the function and get rid of `drivers=DRIVER_FUNCS`

GeeTransit replied:
It'll look weird. Plus, it might be useful to outside functions.

kwangzera replied:
how about defining it somewhere not in the middle of functions and also vscode gives me this warning `Dangerous default value DRIVER_FUNCS (builtins.dict) as argumentpylint(dangerous-default-value)`

GeeTransit replied:
lol The warning is because it's a dictionary. I'll move it to just below the imports

kwangzera commented on typeracer_bot.py:
```diff
@@ -1,78 +1,81 @@
 import time
```
lmao i will reformat ur code my way

GeeTransit replied:
lmao i'll make another pull request to restore my way

kwangzera commented on typeracer_bot.py:
```diff
@@ -1,78 +1,81 @@
+    for name, func in drivers.items():
+        with suppress(WebDriverException):
+            return func()
+    raise LookupError("Driver not found: {preferred=!r}")
```
what does `{preferred=!r}` stand for

GeeTransit replied:
(O shoot I forgot the `f` before the string.) The `!r` applies `repr` over the result.

Example:
```py
>>> a = "derp"
>>> print(f"{a}")
derp
>>> print(f"{a!r}")
"derp"
>>> print(f"{a=}")
a=derp
>>> print(f"{a=!r}")
a="derp"
```

kwangzera commented on typeracer_bot.py:
```diff
@@ -1,78 +1,81 @@
             delay = float(input("Set your delay and press enter when the race starts: "))
         except ValueError:
-            print("- default delay time set to 0.1s")
+            print(f"- using previous delay time of {delay}s")
```
I originally have it default to 0.1s but I forgot to push the version where it had `delay = 0.1` after `print("- default delay time set to 0.1s")`

GeeTransit replied:
lol k i'll revert this.

kwangzera commented on typeracer_bot.py:
```diff
@@ -1,78 +1,81 @@
-from selenium import webdriver
-from selenium.common.exceptions import TimeoutException
+from selenium.webdriver import Chrome, Firefox, Safari
+from selenium.webdriver.remote.webdriver import WebDriver
```
why not `from selenium import webdriver`

GeeTransit replied:
Too much code. Compare `webdriver.Chrome` with `Chrome`. It gets worse with `webdriver.remove.webdriver.WebDriver` vs `WebDriver`. (webdriver gets typed 3 times...)

kwangzera replied:
Won't be needing line 8 (this line) if I do `if not driver` on line 61. Also realized that `webdriver` and `WebDriver` are different things

Togohogo1 commented on typeracer_bot.py:
```diff
@@ -1,78 +1,81 @@
+)
+
+def get_driver(preferred=None, drivers=DRIVER_FUNCS):
+    """Return first working driver in `drivers` dict."""
+    if preferred is not None:
+        with suppress(WebDriverException):
+            return drivers[preferred]()
+    for name, func in drivers.items():
+        with suppress(WebDriverException):
+            return func()
+    raise LookupError("Driver not found: {preferred=!r}")
```
lol clarify

Togohogo1 commented on typeracer_bot.py:
```diff
@@ -1,78 +1,81 @@
+    raise LookupError("Driver not found: {preferred=!r}")
+
+def main(link="https://play.typeracer.com", delay=0.1, driver=None):
+    if not isinstance(driver, WebDriver):
+        driver = get_driver(preferred=driver)
+    keyboard = Controller()
```
lol clarify

kwangzera commented on typeracer_bot.py:
```diff
@@ -1,78 +1,81 @@
+    raise LookupError("Driver not found: {preferred=!r}")
+
+def main(link="https://play.typeracer.com", delay=0.1, driver=None):
+    if not isinstance(driver, WebDriver):
```
can't u just do `if not driver`

GeeTransit replied:
`driver` can be a string (such as `"chrome"`).

GeeTransit replied:
`driver` can have 3 possible types: `None`, `str`, or `WebDriver`.

When it's `None`, it'll go through the drivers like normal (checking if it works). `str` also sends the preferred driver to `get_driver`, meaning it'll attempt that driver first, then continue just like `None`. When a `WebDriver` is passed, it uses it directly, skipping the `get_driver`.

The code needs to check if the driver is a WebDriver (to skip or not to skip), so using `not isinstance(driver, WebDriver)` would work. If the check used `not driver`, a string would be used as the driver (which doesn't work).

kwangzera commented on typeracer_bot.py:
```diff
@@ -1,78 +1,81 @@
+
+def get_driver(preferred=None, drivers=DRIVER_FUNCS):
+    """Return first working driver in `drivers` dict."""
+    if preferred is not None:
```
why not just `if preferred`

GeeTransit replied:
That works too, but I find the complementing `preferred=None` and `preferred is not None` nice to see (and to understand).

GeeTransit committed:
Add missing f in f-string

GeeTransit committed:
Rename main(delay -> default) and change default

Argument `delay` -> `default`
Empty / invalid delay will use default delay (instead of previous)

kwangzera commented on typeracer_bot.py:
```diff
@@ -1,78 +1,82 @@
+    if preferred is not None:
+        with suppress(WebDriverException):
+            return drivers[preferred]()
+    for name, func in drivers.items():
```
how about `for _, func in drivers.items():`

kwangzera replied:
Better idea: `for func in drivers.values():`

GeeTransit replied:
Ye that was left over from when I was gonna raise an exception using the name of the driver.

GeeTransit committed:
Move constant DRIVER_FUNCS to the top

GeeTransit committed:
Make get_driver return name and driver instance

get_driver returns a namedtuple DriverInfo(name, driver)
get_driver now allows a preferred argument not in the drivers dict (checked before getting)

GeeTransit committed:
Remove Safari

GeeTransit committed:
Add Edge support

GeeTransit committed:
Avoid pylint warning for dictionaries in default arguments

GeeTransit committed:
Remove get_driver's namedtuple return value
